zigzag/classes/stages/SearchNoUseMemStage.py

```diff
@@ -0,0 +1,277 @@
+## This stage must be processed before WorkloadStage.
+## This stage figures out the no-use top memory levels for "I", "W", "O" when the size of lower memory level is enough to hold all data, considering the output data of previous layer can be directly used by next layer. As an impact, the energy / latency related to these memories will be removed.
+## The general criteria is:
+##      If a low-level memory size is big enough to hold both "I" and "O" data of current layer, memory above this one will be labeled as no-use.
+##      If a low-level memory size is big enough to hold "W" data of entire workload, memory above this one will be labeled as no-use.
+## The above method only applies layers along the same branch, otherwise (for branch starting nodes or branch final nodes) the "O" data will return back to the top possible memory.
+## In RemoveNoUseMemStage, no-use mem across all layers, labeled in this stage, will be removed in the memory architecture.
+## For now, the number of cores must be 1.
+#################### Pseudo-code ####################
+## Initialization:
+##   mem_update_list = [layer_ids: {"I" / "O": -1}] ## mem level of different operands of each layer (there should be no -1 after self.update_top_mem_level())
+##   each_layer_IO_data_size = [layer_ids: {"I" / "O": size}] ## input / output data size of each layer
+##   mem_update_weight = top_mem_level ## top mem level to put weight
+##   weight_size_entire_workload = weight_size # weight data size of entire workload
+## Generate:
+##   layer_execution_order = list( topological_sort(layer_gragh) )
+## Locate top mem level for each operand of each layer. Store results in mem_update_list and mem_update_weight.
+##   for layer in all_layers:
+##     if layer.index != 0: ## not the 1st execution layer
+##       mem_udpate_list[layer]["I"] = mem_udpate_list[previous_layer]["O"]
+##     if len(layer.next_node) > 1 or len(next_layer.prevous_node) > 1: ## starting node of branches / final node of branches
+##     | if layer.index == 0:
+##     |   mem_update_list[layer]["I" / "O"] updates to the top input/output mem level
+##     | else:
+##     |   mem_update_list[layer]["O"] updates to the top output mem level
+##     |   mem_update_weight = top weight mem level, if mem_update_weight > top weight mem level
+##     |
+##     else:
+##       for mem in mem_levels(sort_order: from top to bottom):
+##         if sum(layer[operand_size] for operand in mem.operands) <= mem.size:
+##           if ["I", "O"] both in mem.operands:
+##             mem_update_list[layer]["O"] = current_mem_level
+##             if layer.index == 0: ## the 1st execution layer
+##               mem_update_list[layer]["I"] = current_mem_level
+##           if ("W" in mem.operand) and (current_mem_level < mem_update_weight):
+##             mem_update_weight = current_mem_level
+#####################################################
+
+class SearchNoUseMemStage(Stage):
```

Comment:
Rename this class and documentation to "SearchUnusedMemoryStage" or something similar.